# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [2]:
import pandas as pd

In [30]:
df = pd.read_csv('horse_data.csv')
df.columns = ['Surgery','Age','HospitalNumber','RectalTemperature','Pulse','RespiratoryRate','TemperatureOfExtremities',
              'PeripheralPulse','mucous membranes','capillary refill time','pain','peristalsis','abdominal distension',
             'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination', 'abdomen',
             'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein',
             'outcome', 'surgical lesion', 'type of lesion1','type of lesion2','type of lesion3', 'cp_data']
df.head(5)

Surgery  Age  HospitalNumber RectalTemperature Pulse RespiratoryRate  \
0       1    1          534817              39.2    88              20   
1       2    1          530334             38.30    40              24   
2       1    9         5290409             39.10   164              84   
3       2    1          530255             37.30   104              35   
4       2    1          528355                 ?     ?               ?   

  TemperatureOfExtremities PeripheralPulse mucous membranes  \
0                        ?               ?                4   
1                        1               1                3   
2                        4               1                6   
3                        ?               ?                6   
4                        2               1                3   

  capillary refill time  ... packed cell volume total protein  \
0                     1  ...                 50            85   
1                     1  ...              33.00          6.70   
2                     2  ...              48.00          7.20   
3                     2  ...              74.00          7.40   
4                     1  ...                  ?             ?   

  abdominocentesis appearance abdomcentesis total protein outcome  \
0                           2                           2       3   
1                           ?                           ?       1   
2                           3                        5.30       2   
3                           ?                           ?       2   
4                           ?                           ?       1   

  surgical lesion type of lesion1 type of lesion2 type of lesion3 cp_data  
0               2            2208               0               0       2  
1               2               0               0               0       1  
2               1            2208               0               0       1  
3               2            4300               0               0       2  
4               2               0               0               0       2  

[5 rows x 28 columns]

In [31]:
df = df.drop(['Age','nasogastric reflux PH','HospitalNumber','PeripheralPulse','TemperatureOfExtremities','peristalsis','nasogastric tube', 'nasogastric reflux','rectal examination', 'abdomen','packed cell volume', 'total protein','abdominocentesis appearance', 'abdomcentesis total protein','outcome', 'surgical lesion', 'type of lesion1','type of lesion2','type of lesion3', 'cp_data'], axis=1)
df.head(15)

Surgery RectalTemperature Pulse RespiratoryRate mucous membranes  \
0        1              39.2    88              20                4   
1        2             38.30    40              24                3   
2        1             39.10   164              84                6   
3        2             37.30   104              35                6   
4        2                 ?     ?               ?                3   
5        1             37.90    48              16                1   
6        1                 ?    60               ?                ?   
7        2                 ?    80              36                3   
8        2             38.30    90               ?                1   
9        1             38.10    66              12                5   
10       2             39.10    72              52                2   
11       1             37.20    42              12                1   
12       2             38.00    92              28                2   
13       1              38.2    76              28                1   
14       1             37.60    96              48                4   

   capillary refill time pain abdominal distension  
0                      1    3                    2  
1                      1    3                    1  
2                      2    2                    4  
3                      2    ?                    ?  
4                      1    2                    2  
5                      1    3                    3  
6                      1    ?                    2  
7                      1    4                    4  
8                      1    5                    1  
9                      1    3                    1  
10                     1    2                    2  
11                     1    3                    3  
12                     1    1                    2  
13                     1    3                    1  
14                     1    5                    3

In [32]:
for col in df.columns:
    df[col] = df[col].apply(lambda x: None if x == '?' else float(x)) 

In [33]:
df.head(15)

Surgery  RectalTemperature  Pulse  RespiratoryRate  mucous membranes  \
0       1.0               39.2   88.0             20.0               4.0   
1       2.0               38.3   40.0             24.0               3.0   
2       1.0               39.1  164.0             84.0               6.0   
3       2.0               37.3  104.0             35.0               6.0   
4       2.0                NaN    NaN              NaN               3.0   
5       1.0               37.9   48.0             16.0               1.0   
6       1.0                NaN   60.0              NaN               NaN   
7       2.0                NaN   80.0             36.0               3.0   
8       2.0               38.3   90.0              NaN               1.0   
9       1.0               38.1   66.0             12.0               5.0   
10      2.0               39.1   72.0             52.0               2.0   
11      1.0               37.2   42.0             12.0               1.0   
12      2.0               38.0   92.0             28.0               2.0   
13      1.0               38.2   76.0             28.0               1.0   
14      1.0               37.6   96.0             48.0               4.0   

    capillary refill time  pain  abdominal distension  
0                     1.0   3.0                   2.0  
1                     1.0   3.0                   1.0  
2                     2.0   2.0                   4.0  
3                     2.0   NaN                   NaN  
4                     1.0   2.0                   2.0  
5                     1.0   3.0                   3.0  
6                     1.0   NaN                   2.0  
7                     1.0   4.0                   4.0  
8                     1.0   5.0                   1.0  
9                     1.0   3.0                   1.0  
10                    1.0   2.0                   2.0  
11                    1.0   3.0                   3.0  
12                    1.0   1.0                   2.0  
13                    1.0   3.0                   1.0  
14                    1.0   5.0                   3.0

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Surgery                298 non-null    float64
 1   RectalTemperature      239 non-null    float64
 2   Pulse                  275 non-null    float64
 3   RespiratoryRate        241 non-null    float64
 4   mucous membranes       253 non-null    float64
 5   capillary refill time  267 non-null    float64
 6   pain                   244 non-null    float64
 7   abdominal distension   243 non-null    float64
dtypes: float64(8)
memory usage: 18.8 KB


In [36]:
df.describe()

Surgery  RectalTemperature       Pulse  RespiratoryRate  \
count  298.000000         239.000000  275.000000       241.000000   
mean     1.395973          38.166527   71.934545        30.427386   
std      0.489881           0.733508   28.680522        17.678256   
min      1.000000          35.400000   30.000000         8.000000   
25%      1.000000          37.800000   48.000000        18.000000   
50%      1.000000          38.200000   64.000000        24.000000   
75%      2.000000          38.500000   88.000000        36.000000   
max      2.000000          40.800000  184.000000        96.000000   

       mucous membranes  capillary refill time        pain  \
count        253.000000             267.000000  244.000000   
mean           2.853755               1.303371    2.942623   
std            1.620294               0.476623    1.303993   
min            1.000000               1.000000    1.000000   
25%            1.000000               1.000000    2.000000   
50%            3.000000               1.000000    3.000000   
75%            4.000000               2.000000    4.000000   
max            6.000000               3.000000    5.000000   

       abdominal distension  
count            243.000000  
mean               2.259259  
std                1.061472  
min                1.000000  
25%                1.000000  
50%                2.000000  
75%                3.000000  
max                4.000000

In [54]:
print('мода Операции:', df['Surgery'].mode()[0])
print('мода Слизистой оболочки:', df['mucous membranes'].mode()[0])
print('мода Боли:', df['pain'].mode()[0])
print('мода Вздутия живота:', df['abdominal distension'].mode()[0])

мода Операции: 1.0
мода Слизистой оболочки: 1.0
мода Боли: 3.0
мода Вздутия живота: 1.0


Surgery - в исследовании большнство лошадей лечились с опрацией. Данный показатель в выборке наиболее полный

RectalTemperature - большинство лошадей имеют повышенную ректальную температуру, только 25% имеют температуру ниже или равную норме 37.8, максимлаьная очень высокая 40.8, минимальная 35.4

Pulse - большинство лошадей имеют прульс выше нормы 30-40, максимлаьный пульс очень высокий 184

RespiratoryRate - среднее значение частоты дыхания очень высокое 30.43, хотя норма 8-10

mucous membranes - чаще всего у лошадей нормальный розовый цвет слизистой оболочки

capillary refill time - в 75% случаев время наполнения капиляров меньше 3 секунд

pain - чаще всего лошади испытывают периодическую легку боль

abdominal distension - важный показатель болезни, чаще всего лошади не имеют вздутия живота

**Выбросы**

In [63]:
q1 = df.RectalTemperature.quantile(0.25)
q3 = df.RectalTemperature.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
print(f'нижняя гранца выброса: {lower_bound}')
print(f'верхняя гранца выброса: {upper_bound}')
upper_outliers = df[df.RectalTemperature > upper_bound].sort_values('RectalTemperature')
lower_outliers = df[df.RectalTemperature < lower_bound].sort_values('RectalTemperature')

нижняя гранца выброса: 36.74999999999999
верхняя гранца выброса: 39.550000000000004


In [60]:
print('Выбросы верхней границы:')
upper_outliers

Выбросы верхней границы:


Surgery  RectalTemperature  Pulse  RespiratoryRate  mucous membranes  \
98       2.0               39.6  108.0             51.0               6.0   
74       1.0               39.7  100.0              NaN               5.0   
19       1.0               39.9   72.0             60.0               5.0   
280      2.0               40.0   78.0              NaN               5.0   
53       2.0               40.3  114.0             36.0               1.0   
90       2.0               40.3  114.0             36.0               1.0   
258      1.0               40.8   72.0             42.0               1.0   

     capillary refill time  pain  abdominal distension  
98                     2.0   2.0                   3.0  
74                     2.0   2.0                   NaN  
19                     2.0   5.0                   4.0  
280                    1.0   2.0                   1.0  
53                     2.0   2.0                   3.0  
90                     2.0   2.0                   3.0  
258                    1.0   2.0                   1.0

In [62]:
print('Выбросы нижней границы:')
lower_outliers

Выбросы нижней границы:


Surgery  RectalTemperature  Pulse  RespiratoryRate  mucous membranes  \
43       1.0               35.4  140.0             24.0               4.0   
140      2.0               36.0   42.0             30.0               5.0   
237      2.0               36.1   88.0              NaN               3.0   
79       1.0               36.4   98.0             35.0               4.0   
117      1.0               36.5   78.0             30.0               1.0   
297      1.0               36.5  100.0             24.0               3.0   
250      2.0               36.6   42.0             18.0               2.0   

     capillary refill time  pain  abdominal distension  
43                     2.0   4.0                   NaN  
140                    1.0   NaN                   NaN  
237                    1.0   3.0                   2.0  
79                     1.0   4.0                   2.0  
117                    1.0   5.0                   1.0  
297                    1.0   3.0                   3.0  
250                    1.0   1.0                   1.0

Выводы по выбросам ректальной температуры: являются частью исследуемой выборки и не подлежат исключению. Лошади могли иметь в критичиеских ситуациях как очень высокую температуру, так и досточной низкую. Значения выбросов не являются аномалией и имеют место быть.

In [64]:
q1 = df.Pulse.quantile(0.25)
q3 = df.Pulse.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
print(f'нижняя гранца выброса: {lower_bound}')
print(f'верхняя гранца выброса: {upper_bound}')
upper_outliers = df[df.Pulse > upper_bound].sort_values('Pulse')
lower_outliers = df[df.Pulse < lower_bound].sort_values('Pulse')

нижняя гранца выброса: -12.0
верхняя гранца выброса: 148.0


In [65]:
print('Выбросы верхней границы:')
upper_outliers

Выбросы верхней границы:


Surgery  RectalTemperature  Pulse  RespiratoryRate  mucous membranes  \
40       2.0               39.0  150.0             72.0               NaN   
274      1.0               38.8  150.0             50.0               6.0   
54       1.0               38.6  160.0             20.0               5.0   
2        1.0               39.1  164.0             84.0               6.0   
254      1.0               38.8  184.0             84.0               1.0   

     capillary refill time  pain  abdominal distension  
40                     NaN   NaN                   NaN  
274                    2.0   5.0                   2.0  
54                     1.0   3.0                   4.0  
2                      2.0   2.0                   4.0  
254                    1.0   4.0                   3.0

In [66]:
print('Выбросы нижней границы:')
lower_outliers

Выбросы нижней границы:


Empty DataFrame
Columns: [Surgery, RectalTemperature, Pulse, RespiratoryRate, mucous membranes, capillary refill time, pain, abdominal distension]
Index: []

Выводы по выбросам Пульса: выбросов по нижней границе выборки не наблюдается. Выбросы верхней границы находятся в зоне экстремальной нагрузки, но в пределах значений, которые в принципе могут быть у лошадей. Максимальное 250 уд/мин. Исходя из этого выброы оставляем для аналитики и не исключаем.

**Пропуски**

In [67]:
for col in df.columns:
    pct_missing = df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

Surgery - 0.3%
RectalTemperature - 20.1%
Pulse - 8.0%
RespiratoryRate - 19.4%
mucous membranes - 15.4%
capillary refill time - 10.7%
pain - 18.4%
abdominal distension - 18.7%


Решение по замене пропусков: все количественные показатели будем заменять медианой, а качественные модой. Удалять значения не будем, т.к. они имеют существенный процент в выборке и без них данные исказятся и будут недостаточными.
Данный подход позволит не искажать первоначальные данные и не сильно повлияет на базовые показатели.

In [69]:
# Surgery - заменим модой
df.Surgery.fillna(df.Surgery.mode()[0], inplace=True)
# RectalTemperature - заменим медианой
df.RectalTemperature.fillna(df.RectalTemperature.median(), inplace=True)
# Pulse  - заменим медианой
df.Pulse.fillna(df.Pulse.median(), inplace=True)
# RespiratoryRate - заменим медианой
df.RespiratoryRate.fillna(df.RespiratoryRate.median(), inplace=True)
# mucous membranes  - заменим модой
df['mucous membranes'].fillna(df['mucous membranes'].median(), inplace=True)
# capillary refill time  - заменим медианой
df['capillary refill time'].fillna(df['capillary refill time'].mode()[0], inplace=True)
# pain - заменим модой
df['pain'].fillna(df['pain'].mode()[0], inplace=True)
# abdominal distension  - заменим модой
df['abdominal distension'].fillna(df['abdominal distension'].mode()[0], inplace=True)

In [71]:
df.head(15)

Surgery  RectalTemperature  Pulse  RespiratoryRate  mucous membranes  \
0       1.0               39.2   88.0             20.0               4.0   
1       2.0               38.3   40.0             24.0               3.0   
2       1.0               39.1  164.0             84.0               6.0   
3       2.0               37.3  104.0             35.0               6.0   
4       2.0               38.2   64.0             24.0               3.0   
5       1.0               37.9   48.0             16.0               1.0   
6       1.0               38.2   60.0             24.0               3.0   
7       2.0               38.2   80.0             36.0               3.0   
8       2.0               38.3   90.0             24.0               1.0   
9       1.0               38.1   66.0             12.0               5.0   
10      2.0               39.1   72.0             52.0               2.0   
11      1.0               37.2   42.0             12.0               1.0   
12      2.0               38.0   92.0             28.0               2.0   
13      1.0               38.2   76.0             28.0               1.0   
14      1.0               37.6   96.0             48.0               4.0   

    capillary refill time  pain  abdominal distension  
0                     1.0   3.0                   2.0  
1                     1.0   3.0                   1.0  
2                     2.0   2.0                   4.0  
3                     2.0   3.0                   1.0  
4                     1.0   2.0                   2.0  
5                     1.0   3.0                   3.0  
6                     1.0   3.0                   2.0  
7                     1.0   4.0                   4.0  
8                     1.0   5.0                   1.0  
9                     1.0   3.0                   1.0  
10                    1.0   2.0                   2.0  
11                    1.0   3.0                   3.0  
12                    1.0   1.0                   2.0  
13                    1.0   3.0                   1.0  
14                    1.0   5.0                   3.0

In [70]:
df.describe()

Surgery  RectalTemperature       Pulse  RespiratoryRate  \
count  299.000000         299.000000  299.000000       299.000000   
mean     1.394649          38.173244   71.297659        29.180602   
std      0.489595           0.655657   27.586016        16.067835   
min      1.000000          35.400000   30.000000         8.000000   
25%      1.000000          37.900000   48.000000        20.000000   
50%      1.000000          38.200000   64.000000        24.000000   
75%      2.000000          38.500000   88.000000        34.500000   
max      2.000000          40.800000  184.000000        96.000000   

       mucous membranes  capillary refill time        pain  \
count        299.000000             299.000000  299.000000   
mean           2.876254               1.270903    2.953177   
std            1.490936               0.460000    1.177736   
min            1.000000               1.000000    1.000000   
25%            1.000000               1.000000    2.000000   
50%            3.000000               1.000000    3.000000   
75%            4.000000               2.000000    4.000000   
max            6.000000               3.000000    5.000000   

       abdominal distension  
count            299.000000  
mean               2.023411  
std                1.075718  
min                1.000000  
25%                1.000000  
50%                2.000000  
75%                3.000000  
max                4.000000

Вывод после замены пропусков:

Все значения изменились несущественно, базовые показали остались почти точно такими же как в изначальном датафрейме.


## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам
Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода.